src/main/java/org/apache/brooklyn/cloudfoundry/location/CloudFoundryPaasLocation.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.brooklyn.cloudfoundry.location;
+
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URL;
+
+import org.apache.brooklyn.config.ConfigKey;
+import org.apache.brooklyn.core.config.ConfigKeys;
+import org.apache.brooklyn.core.location.AbstractLocation;
+import org.apache.brooklyn.location.paas.PaasLocation;
+import org.cloudfoundry.client.lib.CloudCredentials;
+import org.cloudfoundry.client.lib.CloudFoundryClient;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CloudFoundryPaasLocation extends AbstractLocation
+        implements PaasLocation, CloudFoundryPaasLocationConfig {
+
+    public static final Logger log = LoggerFactory.getLogger(CloudFoundryPaasLocation.class);
+
+    public static ConfigKey<String> CF_USER = ConfigKeys.newStringConfigKey("user");
+    public static ConfigKey<String> CF_PASSWORD = ConfigKeys.newStringConfigKey("password");
+    public static ConfigKey<String> CF_ORG = ConfigKeys.newStringConfigKey("org");
+    public static ConfigKey<String> CF_ENDPOINT = ConfigKeys.newStringConfigKey("endpoint");
+    public static ConfigKey<String> CF_SPACE = ConfigKeys.newStringConfigKey("space");
+
+    private CloudFoundryClient client;
+
+    public CloudFoundryPaasLocation() {
```

Comment:
Thanks, just a question. If I add the client as a constructor parameter, how can I create the location. I mean, currently I am using, for example: 
````
     Map<String, String> m = MutableMap.of();
     m.put("user", "super_user");
     m.put("password", "super_secret");
     m.put("org", "secret_organization");
     m.put("endpoint", "https://api.super.secret.io");
     m.put("space", "development");

    return (CloudFoundryPaasLocation)
        mgmt.getLocationRegistry().getLocationManaged("cloudfoundry", m);
````
But now the constructor of (`CloudFoundryPaasLocation`) requires an client. How could I build an location?

I suggest to have some different constructors.
````
    public CloudFoundryPaasLocation(){
        super();
        //the client can not be setUp here because the client requires
        //some `ConfigKeys` and this are not ready on this moment
   }

    public CloudFoundryPaasLocation(Map<?, ?> properties) {
        super(properties);
        client =new CloudFoundryClient(
                       new CloudCredentials(getConfig(CF_USER), getConfig(CF_PASSWORD)),
                       getTargetURL(getConfig(CF_ENDPOINT)), getConfig(CF_ORG), 
                       getConfig(CF_SPACE), true);
        client.login(); //login could be calling from getClient(), lazy login.
    }

    public CloudFoundryPaasLocation(CloudFoundryClient client) {
        super();
        this.client = client;
    }
````
WDYT?